test/inductor/test_templated_attention.py

```diff
@@ -147,23 +157,116 @@ def score_mod_2(score, b, h, m, n):
-    @expectedFailure
     @supported_platform
     @common_utils.parametrize("dtype", test_dtypes)
     def test_captured_buffers(self, dtype: torch.dtype):
-        head_offset = torch.rand(8, device="cuda", dtype=dtype)
+        head_offset = torch.rand(H, device="cuda", dtype=dtype)
 
         def score_mod(score, b, h, m, n):
-            return score + head_offset[h]
+            return score + index(head_offset, [h])
 
         self.run_test(score_mod, dtype)
 
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_seq_masking(self, dtype):
+        seq_idx = torch.zeros(S, device="cuda", dtype=torch.bool)
+        seq_idx[S // 2 :] = 1
+
+        def seq_mask_mod(score, b, h, q, kv):
+            return torch.where(
+                index(seq_idx, [q]) == index(seq_idx, [kv]), score, float("-inf")
+            )
+
+        self.run_test(seq_mask_mod, dtype)
+
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_load_from_bias_seq_only(self, dtype):
+        bias = torch.randn(S, S, device="cuda", dtype=dtype)
+
+        def bias_mod(score, b, h, q, kv):
+            return score + index(bias, [q, kv])
+
+        self.run_test(bias_mod, dtype)
+
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_load_from_bias_seq_batch(self, dtype):
+        bias = torch.randn(B, S, S, device="cuda", dtype=dtype)
+
+        def bias_mod(score, b, h, q, kv):
+            return score + index(bias, [b, q, kv])
+
+        self.run_test(bias_mod, dtype)
+
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_load_from_bias_head_seq_batch(self, dtype):
+        bias = torch.randn(B, H, S, S, device="cuda", dtype=dtype)
+
+        def bias_mod(score, b, h, q, kv):
+            return score + index(bias, [b, h, q, kv])
+
+        self.run_test(bias_mod, dtype)
+
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_load_rel_bias(self, dtype):
+        rel_bias = torch.randn(2 * S, device="cuda", dtype=dtype)
+
+        def bias_mod(score, b, h, q, kv):
+            return score + index(rel_bias, [(q - kv) + S])
+
+        self.run_test(bias_mod, dtype)
+
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    def test_dependent_causal_bidirectional(self, dtype):
+        num_bidirectional = torch.randint(0, S, (B,), device="cuda", dtype=torch.int32)
+
+        def bias_mod(score, b, h, q, kv):
+            causal_attention = q >= kv
+            cur_num_bidirectional = index(num_bidirectional, (b,))
+            bidirectional_attention_on_video = (q <= cur_num_bidirectional) & (
+                kv <= cur_num_bidirectional
+            )
+            return torch.where(
+                bidirectional_attention_on_video | causal_attention,
```

Comment:
could you also maybe try this with composition function?

Reply:
Composition here is actually a little bit tricky. Since you want the two masks to be "or"ed together and not "and"ed.